### PR TITLE
Separating errorSanitizer logic to parse runtime errors also

### LINKED ISSUE #505: Error message rewriting is lost for runtime errors.
We've lost error message rewriting in runtime errors, as part of migrating the rewriting code to the server.  For example, when I lose at https://code.world/#PWCzFEaAgOuBeBUKcEPiVzw, I get:

    GAME OVER
    CallStack (from HasCallStack):
      error, called at Line 34, Column 12 in main:Main

There used to be rules that would replace "main:Main" with "your program", and also "CallStack (from HasCallStack):" with "When evaluating:".  Now that those are bundled into the compiler, they do not get applied to runtime errors. Oops!

The solution is going to be a little ugly.  Perhaps it's to move error sanitizing into a runtime library, which can be compiled by both GHCJS and GHC?

### PR DESCRIPTION
**Pull request by pranjaltale16**
fix #505

### CODE REVIEW

**Comment by pranjaltale16** [comment]
I'm getting some errors with sanitizing. I added functions for that, build is running perfectly but messages are not getting sanitized. 
@cdsmith could you please help me with this, where am I going wrong?

File: build.sh [hunk 1]
```diff
@@ -23,9 +23,10 @@ run .  cabal update
 # Install the codeworld-base and codeworld-api packages
 
 run .  cabal_install --ghcjs ./codeworld-prediction \
+                             ./codeworld-error-sanitizer \
                              ./codeworld-api \
                              ./codeworld-base \
-                             ./codeworld-game-api
+                             ./codeworld-game-api
```
Comment (cdsmith) [inline build.sh]: [nit] No need for the trailing space.

File: codeworld-api/codeworld-api.cabal [hunk 2]
```diff
@@ -27,6 +27,8 @@ Library
                        CodeWorld.Driver
                        CodeWorld.CollaborationUI
   Build-depends:       base                 >= 4.8 && < 5,
+                       bytestring,
```
Comment (cdsmith) [inline codeworld-api/codeworld-api.cabal]: Can you add version bounds for bytestring, please?

File: codeworld-api/codeworld-api.cabal [hunk 3]
```diff
@@ -27,6 +27,8 @@ Library
...
+                       codeworld-error-sanitizer,
```
Comment (cdsmith) [inline codeworld-api/codeworld-api.cabal]: This should be either here or in the impl(ghcjs) section, not both.  (Although, see the other comment about rearranging dependencies)

File: codeworld-api/src/CodeWorld.hs [hunk 4]
```diff
@@ -121,7 +121,8 @@ module CodeWorld (
     pattern PointerMovement,
 
     -- * Debugging
-    trace
+    trace,
+    addMessage
```
Comment (cdsmith) [inline codeworld-api/src/CodeWorld.hs]: Can you add a CodeWorld.Internal package, and only export from that?  I don't really want this to appear as part of the public CodeWorld API.

File: codeworld-api/src/CodeWorld/Driver.hs [hunk 5]
```diff
@@ -1568,6 +1571,9 @@ trace msg x = unsafePerformIO $ do
 reportError :: SomeException -> IO ()
 reportError e = js_reportRuntimeError True (textToJSString (pack (show e)))
 
+addMessage :: SomeException -> IO ()
+addMessage e = js_reportRuntimeError True (textToJSString (pack (C.unpack (filterOutput (C.pack (show e))))))
```
Comment (cdsmith) [inline codeworld-api/src/CodeWorld/Driver.hs]: Actually, this isn't so much what I had in mind.  I was thinking of something more like:

In `codeworld-api:CodeWorld.Internal`:

    addMessage :: Text -> IO ()

    run :: (Text -> IO ()) -> normal parameters -> IO ()
    run messageHandler params = stuff `catch` handle
      where handle exc = do
                messageHandler (pack (show e))
                setErrorState

In the `codeworld-api:CodeWorld`, things like:

    interactionOf :: normal parameters -> IO ()
    interactionOf params = run addMessage params

And in `codeworld-base`:

    interactionOf params = CodeWorld.Internal.run messageHandler params
      where messageHandler = addMessage . sanitizeMessage

Then ErrorSanitizer would be a dependency of `codeworld-base`, not `codeworld-api`.

**Pull request closed without merging**